Initial DynamicDeps support. (#201)
scrapinghub/scrapy-poet: docs/dynamic-deps.rst

.. _dynamic-deps:

====================
Dynamic dependencies
====================

Normally the dependencies for a callback are specified statically, as type
hints for its arguments:

.. code-block:: python

    import scrapy

    class BooksSpider(scrapy.Spider):
        ...

        def start_requests(self):
            yield scrapy.Request("http://books.toscrape.com/", self.parse_book)

        def parse_book(self, response, book_page: BookPage, other_dep: OtherDep):
            ...

In some cases some or all of the dependencies need to be specified dynamically

> instead, e.g. because they need to be different for different pages using the

Reviewer: ```suggestion
instead, e.g. because they need to be different for different requests using the
```

"pages" looks confusing, as reader may understand it as "for different page objects"

same callback. You can use :class:`scrapy_poet.DynamicDeps
<scrapy_poet.injection.DynamicDeps>` for this. If you add a callback argument
with this type you can pass a list of additional dependency types in the
request meta dictionary using the "inject" key:

.. code-block:: python

    import scrapy

    class BooksSpider(scrapy.Spider):
        ...

        def start_requests(self):
            yield scrapy.Request(
                "http://books.toscrape.com/",
                self.parse_book,
                meta={"inject": [OtherDep]},
            )

        def parse_book(self, response, book_page: BookPage, dynamic: DynamicDeps):

>             other_dep = dynamic[OtherDep]

Reviewer: Is there a way to get a list of the created dependencies in the DynamicDeps (i.e. the available keys)? It seems we should document it. Currently only getting by type is documented, if I'm not mistaken.

Author: I assumed the user already knows which deps are there because it's they who set the meta key, but in other uses cases (which?) they can look at `dynamic.keys()`, do you want to document this or are you thinking about something else?

            ...

> The types passed this way will be used in the dependency resolution as usual,
> and the created instances will be available in the

Reviewer: ```suggestion
The types passed this way are used in the dependency resolution as usual,
with the created instances available in the
```
